### DISCUSSION TITLE
SSG: when is it reasonable to do client-side pagination?

### DISCUSSION THREAD

**Discussion by adamalfredsson** [vercel/next.js#38364] [Help] [answered]
I'm building a statically generated Next.js blog which fetch data from a CMS. In our CMS we have a lot of posts. On our posts page, we want to show a posts list. We want to show 10 posts with a button to load another 10 posts. We also want to filter posts by category.

I see two ways to go about this:

1. We could fetch all posts at build-time using `getStaticProps` and just toggle client-side how many posts should be displayed on the page, and filter with client-side javascript.

2. Or we could fetch 10 posts at build-time and use something like SWR or react-query to fetch more data from the client.

It seems vastly more simple to fetch everything at build-time, so that we don't have to bother with authorizing client requests to our CMS (or setup proxies etc).

However I also realize that it is at somepoint not reasonable to do this. The question is what the bottle-neck with the first approach is and at what quantity of posts I should lean towards the second option instead?

**Comment by icyJoseph** [comment, marked as answer]
> 1. We could fetch all posts at build-time using getStaticProps and just toggle client-side how many posts should be displayed on the page, and filter with client-side javascript.

If you go over some arbitrary, yet sensible, size, in kB range, around 128 kB, then you'll start to see this warning:

```shell
Warning: data for page “/” is 150 kB, this amount of data can reduce performance.
```

In general because, if you fetch all post at build time, these are shared with the client as JSON, then your client app will take longer to download, parse and execute.

> 2. Or we could fetch 10 posts at build-time and use something like SWR or react-query to fetch more data from the client.

Yes, but then you might lose on `SEO`, it really depends on your pagination requirements.

Or, else, you could create something like, `/posts/[page]`, and then use `getStaticPaths` to get how many pages you need, and render as many.

The call to your CMS in `getStaticProps` should also return how many pages there is, in total, so that each page knows how many to render in a navigation UI element.

For example:

```tsx
// pages/posts/[page].tsx
import type { GetStaticPaths, GetStaticProps, NextPage } from "next";
import Link from "next/link";
import { useRouter } from "next/router";
import { useState } from "react";

export const PostsPage: NextPage<{
  total: number;
  posts: Array<{ id: string; content: string }>;
}> = (props) => {
  const [counter, setCounter] = useState(0);
  const { page } = useRouter().query;
  const { total, posts } = props;

  return (
    <div>
      Post page: {page} from a total: {total}
      <div>
        <button onClick={() => setCounter((x) => x + 1)}>
          Count: {counter}
        </button>
      </div>
      <section>
        <h2>posts list</h2>

        <ul>
          {posts.map((post) => (
            <li key={post.id}>{post.content}</li>
          ))}
        </ul>
      </section>
      <nav>
        <ul>
          {Array.from({ length: total }, (_, i) => {
            // for this demo it is ok to use the index as key
            // but ideally not
            return (
              <li key={i}>
                <Link href={`/posts/${i}`}>
                  <a>Go to {i}</a>
                </Link>
              </li>
            );
          })}
        </ul>
      </nav>
    </div>
  );
};

/* get this from your CMS */
const paths = [
  { params: { page: "0" } },
  { params: { page: "1" } },
  { params: { page: "2" } },
  { params: { page: "3" } },
];

export const getStaticPaths: GetStaticPaths<{ page: string }> = () => {
  return {
    paths,
    fallback: false,
  };
};

export const getStaticProps: GetStaticProps = async (ctx) => {
  const page = ctx.params?.page;

  if (typeof page === "undefined") return { notFound: true };
  if (Array.isArray(page)) return { notFound: true };

  console.log({ page });
  // get this together with all of the posts that go on this page
  const total = await new Promise((next) => next(paths.length));

  return {
    props: {
      posts: Array.from({ length: Number(page) + 1 }, (_, i) => ({
        id: i,
        content: `page: ${page} foo`.repeat(Number(page) + 1),
      })),
      total,
    },
  };
};

export default PostsPage;
```

If you start at `/posts/0`, increase the count by pressing the button, and then navigate to `/posts/3`, the state of the counter persists, but the page is now elsewhere,  you can see the posts changing.

Of course if there's a ton of pages in total, then you might want to use `fallback: true` in `getStaticPaths` to avoid generating them all at build time (then you also need to check `useRouter().isFallback`, https://nextjs.org/docs/api-reference/data-fetching/get-static-paths#fallback-pages). 

> However I also realize that it is at somepoint not reasonable to do this. The question is what the bottle-neck with the first approach is and at what quantity of posts I should lean towards the second option instead?

At 128kB data size per page.

**Reply by adamalfredsson** [reply]
Brilliant, thank you so much!

**Reply by JordanDDisch** [reply]
just to add to this, if you decide to go down the paginations route and not infinite scroll / same page load, then the other scaling issue with SSG is just building all those extra pages. I'm doing this currently for a review site (100s of pages) per product review.
